### ISSUE TITLE
2-way audio with ICSee/dvrip cameras

### ISSUE THREAD

**Issue by felipecrs** [AlexxIT/go2rtc#633]
Many cheap chinese cameras are using this ICSee. I bought this one: https://aliexpress.com/item/1005005731238053.html

It's relatively good, although it does not have the capability of changing the stream codec from H265 to H264.

But I have another ICSee based camera as well.

The app is good as well. A lot better than Tuya.

It looks like these cameras uses the DVRIP protocol. My RTSP URLs looks like:

```
rtsp://192.168.1.21:554/user=tema_password=<password>_channel=0_stream=0&onvif=0.sdp
```

It would be awesome if 2-way audio could work with it.

**Comment by felipecrs** [comment]
2-way audio works through its web UI:

![image](https://github.com/AlexxIT/go2rtc/assets/29582865/114fa912-1da3-4e52-baa5-cce36a587b95)

Although it requires the installation of a "VideoPlayTool" to properly open it.

@AlexxIT if someday you are interested in implementing this I can give you access to the camera.

**Comment by felipecrs** [comment]
Oh, and by the way DVRIP source also works with this camera:

```
streams:
  icsee_camera:
    - dvrip://tema:<password>@192.168.1.21:34567?channel=0&subtype=1
```

**Comment by slydiman** [comment]
Agreed. It would be great to support 2-way audio for DVRIP/XM/Sofia-based cameras.

**Comment by slydiman** [comment]
https://github.com/OpenIPC/camerasrnd contains links to NetSDK. The main code is compiled to NetSdk.dll and NetSdk.lib.
ClientDemo sources contain the code like
```
// can set the talk format, default is G711A.
H264_DVR_API bool CALL_METHOD H264_DVR_SetTalkMode(long lLoginID, SDK_AudioInFormatConfig* pTalkMode);

//sending talk data
BOOL CDeviceEditDlg::SendTalkData(LPBYTE pDataBuffer,DWORD dwDataLength)
{
	if ( NULL == m_pTalkDecodeBuf )
	{
		m_pTalkDecodeBuf = new BYTE[4096];
	}
	memset(m_pTalkDecodeBuf, 0, 4096);
	
	int iCbLen = 0;
	
	//encode talk data
	if (g711a_Encode( (char*)pDataBuffer , (char*)m_pTalkDecodeBuf+8 , dwDataLength, &iCbLen) != 1)
	{
		ASSERT(FALSE);
		return FALSE;
	}
	
	m_pTalkDecodeBuf[0]=0x00;
	m_pTalkDecodeBuf[1]=0x00;
	m_pTalkDecodeBuf[2]=0x01;
	m_pTalkDecodeBuf[3]=0xFA;
	
	m_pTalkDecodeBuf[4]=0x0E; //G711A
	m_pTalkDecodeBuf[5]=0x02;
	m_pTalkDecodeBuf[6]=BYTE(iCbLen&0xff);
	m_pTalkDecodeBuf[7]=BYTE(iCbLen>>8);	
	
	iCbLen += 8;
	return H264_DVR_VoiceComSendData( m_lTalkHandle, (char*)m_pTalkDecodeBuf, iCbLen );
}
```
It is possible to disassemble NetSdk.dll functions like H264_DVR_StartVoiceCom_NVR_IPC, H264_DVR_VoiceComSendData, H264_DVR_StopVoiceCom, H264_DVR_SetTalkMode to figure out the exact messages for playback audio on the camera.
Hope this may help.

**Comment by slydiman** [comment]
Here is the `Talk` command implementation for DVRIP protocol
https://github.com/667bdrm/sofiactl/blob/master/sofiactl.pl#L2286
The source file must be PCM G711 ALAW 8kHz mono.
```
    $decoded = $dvr->PrepareGenericCommand(IPcam::TALK_CLAIM, { # 1434
        Name => 'OPTalk',
        OPTalk => {
            Action => "Claim",
            AudioFormat => {
                BitRate => 519788,
                EncodeType => "G711_ALAW",
                SampleBit => 8,
                SampleRate => 8
            }
        }
    });

    $decoded = $dvr->PrepareGenericCommand(IPcam::TALK_REQ, { # 1430
        Name => 'OPTalk',
        OPTalk => {
            Action => "Start",
            AudioFormat => {
                BitRate => 128,
                EncodeType => "G711_ALAW",
                SampleBit => 8,
                SampleRate => 8000
            }
        }
    });

    if ($cfgInputFile ne '') {
        my $size = -s $cfgInputFile;
        print "File size: $size\n";
        my $count = 0;
        open(IN, "< $cfgInputFile");
        while ($count < $size) {
            my $data;
            read(IN, $data, 320);
            #print $data . "\n";
            $count += 320;
            my $audio_pkt = $dvr->BuildRawPacket(IPcam::TALK_CU_PU_DATA, # 1432
                pack('CCCCCCS', (0x00, 0x00, 0x01, 0xFA, 0x0E, 0x02, 320)) . $data);
            $dvr->{socket}->send($audio_pkt);
            my $reply_head = $dvr->GetReplyHead();
            $dvr->GetReplyData($reply_head);
        }
        close(IN);
    }
```

**Comment by AlexxIT** [comment]
Supported in latest master
```yaml
streams:
  stream: dvrip://test:test@b3.slydiman.me:34569
  tts_only: dvrip://test:test@b3.slydiman.me:34569?backchannel=1
  two_way:
    - dvrip://test:test@b3.slydiman.me:34569
    - dvrip://test:test@b3.slydiman.me:34569?backchannel=1
```

**Comment by AlexxIT** [comment]
Supported in latest master:
```yaml
streams:
  stream: dvrip://username:password@192.168.1.123:34567
  tts_only: dvrip://username:password@192.168.1.123:34567?backchannel=1
  two_way:
    - dvrip://username:password@192.168.1.123:34567
    - dvrip://username:password@192.168.1.123:34567?backchannel=1
```

**Comment by felipecrs** [comment]
OMG this is amazing. I will test it asap.

**Comment by Djeloan** [comment]
> Supported in latest master:
> 


Means that will be available in some next update or the current one (1.7.1) ?
Tested it now, but couldn't get it to work, so im guessing it's in the next update

**Comment by AlexxIT** [comment]
@felipecrs have you tried?

**Comment by slydiman** [comment]
Great work! Thank you very much!
But it works with the microphone only from go2rtc addon UI: Links, select `video+audio+microphone = two way audio from camera`, click `webrtc.html` local WebRTC viewer.
I was shocked - there is no way to turn on the microphone in the HA UI. The solution with browser_mod does not work for me (in case of few HA users and Android app).
The next step is to ubdate WebRTC camera custom card to enable the microphone.
Install the Frigate/Frigate addon + Frigate integration + Frigate custom card for 1 button `microphone` is overkill!

**Comment by felipecrs** [comment]
Frigate Card is now supporting external go2rtc in latest dev. Meaning you don't need Frigate.

**Comment by felipecrs** [comment]
> @felipecrs have you tried?

I'm sorry not yet. But I will try it very soon. My ICSee camera is currently far away from my current location.

**Comment by AlexxIT** [comment]
@felipecrs you can send audio to it an check result via stream

**Comment by felipecrs** [comment]
Oh, yes. This is what I see:

```jsonc
{
  "producers": [
    {
      "type": "RTSP passive producer",
      "url": "rtsp://127.0.0.1:8554/8aa14331e4a4907604463ca9c8ca54eb",
      "remote_addr": "127.0.0.1:38036",
      "user_agent": "ffmpeg/go2rtc",
      "sdp": "v=0\r\no=- 0 0 IN IP4 127.0.0.1\r\ns=go2rtc/1.7.1\r\nc=IN IP4 127.0.0.1\r\nt=0 0\r\na=tool:libavformat LIBAVFORMAT_VERSION\r\nm=video 0 RTP/AVP 96\r\na=rtpmap:96 H264/90000\r\na=fmtp:96 packetization-mode=1; sprop-parameter-sets=Z2QMKawsqAoBbpp4AAADAAgAAAMAxHhAIVA=,aO44sA==; profile-level-id=640C29\r\na=control:streamid=0\r\n",
      "medias": ["video, recvonly, H.264 High 4.1"],
      "receivers": ["96 H264, bytes=5876964, senders=1"],
      "recv": 5928476
    },
    {
      "type": "RTSP active producer",
      "url": "rtsp://127.0.0.1:8554/rua_h265?audio=all",
      "remote_addr": "127.0.0.1:8554",
      "user_agent": "go2rtc/1.7.1",
      "sdp": "v=0\r\no=- 1 1 IN IP4 0.0.0.0\r\ns=go2rtc/1.7.1\r\nc=IN IP4 0.0.0.0\r\nt=0 0\r\nm=audio 0 RTP/AVP 96\r\na=rtpmap:96 PCMA/8000\r\na=control:trackID=0\r\nm=audio 0 RTP/AVP 97\r\na=rtpmap:97 MPEG4-GENERIC/8000\r\na=fmtp:97 profile-level-id=1;mode=AAC-hbr;sizelength=13;indexlength=3;indexdeltalength=3; config=158856E500\r\na=control:trackID=1\r\n",
      "medias": [
        "audio, recvonly, PCMA/8000",
        "audio, recvonly, MPEG4-GENERIC/8000"
      ],
      "receivers": ["96 PCMA/8000, bytes=297920, senders=1"],
      "recv": 453571
    }
  ],
  "consumers": [
    {
      "type": "WebRTC/WebSocket async passive consumer",
      "remote_addr": "udp4 srflx 179.106.189.52:64259 related 192.168.1.15:64259",
      "user_agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/118.0.0.0 Safari/537.36",
      "medias": [
        "video, sendonly, VP8, RTX, VP9, H264, AV1, RED, ULPFEC, FLEXFEC-03",
        "audio, sendonly, OPUS/48000/2, RED/48000/2, G722/8000, PCMU/8000, PCMA/8000, CN/8000, TELEPHONE-EVENT/48000, TELEPHONE-EVENT/8000, L16, PCML",
        "audio, recvonly, OPUS/48000/2, RED/48000/2, G722/8000, PCMU/8000, PCMA/8000, CN/8000, TELEPHONE-EVENT/48000, TELEPHONE-EVENT/8000"
      ],
      "receivers": ["111 OPUS/48000/2, bytes=16351, senders=0"],
      "senders": [
        "102 H264, bytes=1290962, receivers=1",
        "8 PCMA/8000, bytes=65920, receivers=1"
      ],
      "recv": 23261,
      "send": 1371330
    }
  ]
}
```

**Comment by felipecrs** [comment]
Oh sorry, wait a sec. I think I made a mistake.

**Comment by AlexxIT** [comment]
I don't need logs. Just check via live stream

**Comment by felipecrs** [comment]
Yes I can hear myself!!! That's amazing! It's working nice, thanks a lot!

**Comment by AlexxIT** [comment]
Nice. Thanks to @slydiman for perl and sharing access to camera.

**Comment by slydiman** [comment]
Sure. HA with https and go2rts master work just fine. 

I tried to update AlexxIT's WebRTC camera custom card today. I have successfully enabled the microphone recording in the browser but stuck with creating the correct rtc offer request.

**Comment by felipecrs** [comment]
How does your go2rtc URL looks like? Mine is:

https://go2rtc-home.<mydomain>.com/

**Comment by slydiman** [comment]
> Note that you cannot use the hassio ingress for this job.

Why? Please provide more datails. I planned to use go2rtc addon for this purpose. I can configure stand alone go2rtc, but I did not figure out is there any authorization? HA must work with 2 way audio otside the home network.

**Comment by felipecrs** [comment]
You don't need to configure a standalone go2rtc, but you need to have go2rtc running with its own domain (not through hassio ingress), so it can be accessed directly by the card.

You can keep using the add-on, just make sure to configure HTTPS/TLS.

It is a long story as to why this is necessary.

**Comment by slydiman** [comment]
> You can keep using the add-on, just make sure to configure HTTPS/TLS.

How to configure go2rtc addon to run it with its own domain (not through hassio ingress).
Can you provide a config sample?

And Can you provide a config sample for Frigate card w/o Frigate integration. Frigate documentation says that 2 way audio requires a frigate camera provided by the Frigate integration.

**Comment by felipecrs** [comment]
Sorry, I have no config samples to provide. Also, I don't use the Frigate card without the Frigate integration.

> Frigate documentation says that 2 way audio requires a frigate camera provided by the Frigate integration.

I think you missed my point that this feature was only added to the Frigate card development version.

Maybe you can find more help in https://github.com/dermotduffy/frigate-hass-card/issues/1206.

**Comment by slydiman** [comment]
> I think you missed my point

I'm not missing anything. I think here is the reason for your confusion:

> Also, I don't use the Frigate card without the Frigate integration.

The issue you mentioned contains the following quiestion [`Frigate Hass Card without Frigate NVR`](https://github.com/dermotduffy/frigate-hass-card/issues/1243) and the answer is `Hopefully this can be happening someday... Or the webrtc card with your microphone support`

I have configured NGINX to access go2rtc addon through HTTPS port 1985. But go2rtc does not have any authorization and it will be a security hole. I can't use it outside my home network. 

I have added the following
```
        - type: custom:frigate-card
          cameras:
          - camera_entity: camera.some_generic
            live_provider: go2rtc
            go2rtc:
              url: https://mydomain:1985
              stream: cam_with_audio
              modes:
               - webrtc
          menu:
            buttons:
              microphone:
                enabled: true
```
Note frigate-card requires camera_entity anyway and it must be a camera from frigate integration. 
I got the error `Could not get camera endpoint` with this config and `Could not determine suitable engine for camera.` w/o camera_entity. 

The bottom line is 
- frigate-card does not work w/o frigate.
- Frigate addon/integration is overkill.
- It is necessary to update WebRTC Camera card to support microphone.

**Comment by felipecrs** [comment]
> But go2rtc does not have any authorization and it will be a security hole. I can't use it outside my home network.

Yes, that is far from ideal. But you can set go2rtc to require user and password, and maybe you can embed the user and password in the go2rtc.url in the frigate card like `https://user:password@mydomain.com:1935`.

I agree it would be great if WebRTC Camera had built-in support for 2-way audio. But @AlexxIT is aware, he just didn't have the time and to ship it yet.

Similarly, support for go2rtc through the hassio ingress is something @dermotduffy [is also aware and he would like to ship it](https://github.com/dermotduffy/frigate-hass-card/issues/1206#issuecomment-1755562794). Likewise, he probably didn't have time to work on it yet.

Needless to say that both has no obligation to work on anything and probably contributions are welcome.

Few months ago we did not have any kind of 2-way audio support at all in Home Assistant. Things are evolving, and will continue to evolve and be more seamless as the time goes. :)

**Comment by slydiman** [comment]
I have updated WebRTC Camera. 

https://github.com/AlexxIT/WebRTC/pull/579

It works just fine in a browser with go2rtc addon. HA must be HTTPS. 

Unfortunately the mic does not work inside the Android HA app - it shows that the mic is used, but there is no any transmission.

**Comment by AlexxIT** [comment]
https://github.com/AlexxIT/go2rtc/releases/tag/v1.8.0

**Comment by Cr4z33** [comment]
Sorry if my question might sound silly, but is there any chance this can work also with a Amcrest AD410 video doorbell or does it have no DVRIP support?

**Comment by slydiman** [comment]
> Amcrest AD410 video doorbell

Seems it uses VoIP/SIP protocols for audio conversations. It is not DVRIP compatible at all.

**Comment by Cr4z33** [comment]
That's what I was fearing... OK thanks.

**Comment by AlexxIT** [comment]
Amcrest AD410 doorbell support RTSP with two way audio

**Comment by felipecrs** [comment]
Guys, since we are talking about two-way audio here, I have two feature requests in Frigate Card that you may be useful for you too:

- https://github.com/dermotduffy/frigate-hass-card/issues/1130
- https://github.com/dermotduffy/frigate-hass-card/issues/1241

If you find them useful, please vote on them with a thumbs up. But please **do not vote if you don't find them useful**.

**Comment by felipecrs** [comment]
Another one:

- https://github.com/dermotduffy/frigate-hass-card/issues/1299

**Comment by PaulBiod** [comment]
Hi,
I'm also using a chinese icsee cam which supports 2 ways audio, but I can't get it to work (updated to last webrtc.), I have no microphone on the stream.

My card looks like this : 

`type: custom:webrtc-camera
entity: camera.jardin_profile_000
ui: true
media: video,audio,microphone
mic: true
title: Jardin
poster: https://home-assistant.io/images/cast/splash.png
muted: false
intersection: 0.75
background: true
style: 'video {aspect-ratio: 16/9; object-fit: fill;}'
shortcuts:
  - name: Record
    icon: mdi:record-circle-outline
    service: switch.toggle
    service_data:
      entity_id: switch.camera_record
mode: webrtc,mse,hls,mjpeg`

any help would be appreciated

**Comment by felipecrs** [comment]
How does the go2rtc config looks like?

Also is your HA exposed as HTTPS?

**Comment by PaulBiod** [comment]
I use webrtc to show cards of my onvif cams (setup in ICSee)
![image](https://github.com/AlexxIT/go2rtc/assets/28720176/da03b5ab-7446-4fc6-9b22-5b1005ffa89d)

Yes, my HA is exposed as https via duckdns

**Comment by felipecrs** [comment]
You need to add a go2rtc configuration with the drip: back channel link.

**Comment by PaulBiod** [comment]
sorry for beeing noob, but how can I achieve this ? I only install webrtc to have greater cam cards, and go2rtc is included in web2rtc hacs integration

**Comment by felipecrs** [comment]
I use the go2rtc add-on, but I think you can probably create one file named go2rtc.yaml at config with something like:

```yaml
streams:
  camera.jardin_profile_000: 
    - "{input}"
    - dvrip://username:password@192.168.1.123:34567?backchannel=1
```

But I am not sure if the "{input}" line will work as I intended. @AlexxIT can comment more on this.

**Comment by slydiman** [comment]
go2rtc add-on config must look like
```
streams:
  camera.jardin_profile_000: 
   - dvrip://admin:password@192.168.1.123:34567?channel=0&subtype=1
   - dvrip://admin:password@192.168.1.123:34567?backchannel=1
```

> My card looks like this :

Here is the correct sample of webrtc-camera card  https://github.com/AlexxIT/WebRTC/pull/579#issuecomment-1763474759
Forget about `mic: true`. It was my suggestion in the pool request. The final implementation requires `media: video,audio,microphone`.
In your case
```
type: 'custom:webrtc-camera'
ui: true
streams:
- url: camera.jardin_profile_000
  mode: webrtc
  name: '🎙️ Mic on'
  media: video,audio,microphone
muted: false
...
```

**Comment by PaulBiod** [comment]
Hi all,
thanks for your time, so, If I understood, I just have to create a file named go2rtc.yaml in the same path where my HAOS config file is, with just these lines (with my IPs of course and credentials): 

streams:
  camera.jardin_profile_000: 
   - dvrip://admin:password@192.168.1.123:34567?channel=0&subtype=1
   - dvrip://admin:password@192.168.1.123:34567?backchannel=1
   
And then use this code for the card : 

type: 'custom:webrtc-camera'
ui: true
streams:
- url: camera.jardin_profile_000
  mode: webrtc
  name: '🎙️ Mic on'
  media: video,audio,microphone
muted: false
...

Thanks guys

**Comment by slydiman** [comment]
> If I understood, I just have to create a file named go2rtc.yaml in the same path where my HAOS config file is

Right. Be sure the addon go2rtc is started. You can open go2rtc addon settings and click `Open web UI` or enable the option `Show in sidebar`, then click on `go2rtc` in left sidebar. You will see go2rtc's web UI. Then you can click on `Config` tab, edit the config and then click the button `Save & Restart`.

**Comment by PaulBiod** [comment]
Ok guys,
so, I created the file : 
![Capture](https://github.com/AlexxIT/go2rtc/assets/28720176/16be91fb-ccc3-4e08-9faf-7ddffdd5ebfc)

Then, restarted HAOS, went to my dashboard, and create a card : 

![Capture3](https://github.com/AlexxIT/go2rtc/assets/28720176/d838bb8a-8d5a-4ec1-92dd-95ae3e8fbd73)

The result is the same (no micro) : 
![Capture2](https://github.com/AlexxIT/go2rtc/assets/28720176/203eed61-a188-4880-a639-1a1745c5d6cc)

Here is my onvifer result showing the ports : 

![Capture d’écran 2023-10-21 141622](https://github.com/AlexxIT/go2rtc/assets/28720176/230ff7ab-0b28-43c2-93a0-459020097798)

Any idea ?
Thanks

**Comment by slydiman** [comment]
No need to restart HAOS. It is enough to restart only go2rtc addon after changing go2rtc.yaml

If you need to see the name of the stream with the mic icon and be able to switch off the mic just add a second stream with the same camera, other name and `media: video,audio`

Be sure you have set permissions to the browser to use the microphone. You must see the rec icon on the page tab. If you are using Google Chrome
Video with muted audio:
![Screenshot 2023-10-21 180216](https://github.com/AlexxIT/go2rtc/assets/16115622/aa2b74f0-66eb-41e2-8d4d-22657da41009)
Video and audio:
![Screenshot 2023-10-21 180225](https://github.com/AlexxIT/go2rtc/assets/16115622/95403c2b-6b4a-4f47-a46b-fffdbc08cb95)
Video, audio and microphone:
![Screenshot 2023-10-21 180237](https://github.com/AlexxIT/go2rtc/assets/16115622/e5191cd2-3038-4035-89de-c3a6d43db671)
Click on the mic icon in the address bar, allow the mic access and select the audio input (microphone) device:
![Screenshot 2023-10-21 180611](https://github.com/AlexxIT/go2rtc/assets/16115622/3b3e61c4-e70c-4912-875f-e9802aae5850)

**Comment by PaulBiod** [comment]
Thanks, indeed the micro in the adress bar was disabled, I enabled it, refresh the page, put a second stream, now I can switch between the streams, but the stream with micro ON doesn't send any audio from my PC...
![image](https://github.com/AlexxIT/go2rtc/assets/28720176/ff98ded6-caa2-46fb-bacf-e7c6e1b8021d)

![image](https://github.com/AlexxIT/go2rtc/assets/28720176/f3e22cbd-e753-4214-a42e-3d5068bbe665)

**Comment by PaulBiod** [comment]
So, I tried with my tablet, it asked me for audio permissions, then I can hear myself speaking, but very low, I only realised it because of the larsen (being close to the camera). The cam speaker is at 100% volume, when I try with icsee, it's clear and loud.
With my laptop, no sound at all

**Comment by eerison** [comment]
I didn't test it yet, But I want to say thank you guys for all effort in this feature ❤️

**Comment by eerison** [comment]
Hey @dbuezas
I guess you was trying to make this work, don't you?

**Comment by felipecrs** [comment]
It's working just fine for me since the day it was implemented.

**Comment by RCW679** [comment]
> Although it requires the installation of a "VideoPlayTool" to properly open it.

This VideoPlayTool is downloaded over insecure HTTP!

**Comment by fedd** [comment]
> Supported in latest master:
> 
> streams:
>   two_way:
>     - dvrip://username:password@192.168.1.123:34567
>     - dvrip://username:password@192.168.1.123:34567?backchannel=1

Hello, this works strangely in Frigate. When I press the mic button I'm able to talk and voice is transmitted, however the video stops and gets distorted, this is on Chrome browser in Android. When I press the mic button in Linux desktop Firefox, the video disappears completely to the loading wheel.

upd: ended up with this config:

```
ffmpeg:
  hwaccel_args: preset-intel-qsv-h264
  global_args: -hide_banner -loglevel warning -threads 2

.....
go2rtc:
  webrtc:
    candidates:
      - 10.75.6.12:8555
      - stun:8555
  streams:
    stairs_orig:
      - dvrip://admin@10.75.16.39:34567?channel=0&subtype=0
    stairs:
      #- rtsp://admin@10.75.16.39/cam/realmonitor?channel=1&subtype=0
      #- ffmpeg:rtsp://10.75.16.39:554/user=admin_password=tlJwpbo6_channel=0_stream=0&onvif=0.sdp?real_stream#video=h264#audio=aac#hardware
      - "ffmpeg:stairs_orig#video=h264#audio=copy#hardware"
      - dvrip://admin@10.75.16.39:34567?backchannel=1
```

however the N100 of my beelink eq13 seems to fall back to software acceleration, at least Frigate used to report something other then 0% for GPU before I'd started these experiments.

**Comment by RomuloGatto** [comment]
@fedd did you succeed in fixing this video disappearing? I'm having this same issue

**Comment by fedd** [comment]
> @fedd did you succeed in fixing this video disappearing? I'm having this same issue

I think I did, I believe transcoding to h264 helped like in the yaml piece above, as it seemingly switches to RTC when the mic button is pressed, and it's written somewhere that RTC doesn't support h265 which my camera produces.

However it echoes wildly and the audio delay is huge, like half second

**Comment by felipecrs** [comment]
H.265 in WebRTC is supported in Chrome 136+, so that shouldn't be an issue unless you're using an old Chrome.

**Comment by RomuloGatto** [comment]
I'm on chrome, but on my mac. Also, on my iPhone, Safari has the same behaviour.

**Comment by felipecrs** [comment]
What is your go2rtc config, @RomuloGatto?

**Comment by felipecrs** [comment]
This is how I am using it:

```yaml
go2rtc:
  streams:
    rua:
      - rtsp://192.168.1.22/user=admin_password=admin_channel=0_stream=1&onvif=0.sdp?real_stream
      - dvrip://admin:admin@192.168.1.22:34567?backchannel=1
```

Try something similar.

**Comment by RomuloGatto** [comment]
> What is your go2rtc config, [@RomuloGatto](https://github.com/RomuloGatto)?

I tried a lot of stuff, but all keep the same behaviour. My current one is this:

```
go2rtc:
  streams:
    rua:
      - ffmpeg:rtsp://user:pass@192.168.x.x:554/Streaming/Channels/101#video=copy#audio=copy
      - dvrip://user:pass@192.168.x.x:34567?backchannel=1#media=audio
```

After you sent yours, I did this

```
go2rtc:
  streams:
    rua:
      - rtsp://user:pass@192.168.x.x:554/admin_channel=0_stream=1&onvif=0.sdp?real_stream
      - dvrip://user:pass@192.168.x.x:34567?backchannel=1
```

And the behaviour persists, the mic still freezes the video.
I'm using a chinese TpTek, if it helps anything

**Comment by felipecrs** [comment]
The structure of your RTSP URLs seem odd. Maybe try:

```yaml
go2rtc:
  streams:
    rua:
      - dvrip://admin:admin@192.168.1.22:34567
      - dvrip://admin:admin@192.168.1.22:34567?backchannel=1
```

Like @AlexxIT [suggested](https://github.com/AlexxIT/go2rtc/issues/633#issuecomment-1758102833).

**Comment by RomuloGatto** [comment]
Same behaviour. I also found a really odd issue. When I access frigate via my external domain, the mic shows up, but when I access it via ip:5000, it doesn't show. Does it trigger anything in your mind?

Also, that's my probe:
```
{
  "producers": [
    {
      "type": "DVRIP active producer",
      "medias": [
        "video, recvonly, H265",
        "audio, recvonly, PCMA/8000"
      ],
      "receivers": [
        "255 H265, bytes=45118615, senders=2",
        "0 PCMA/8000, bytes=2258560, senders=1"
      ]
    },
    {
      "type": "DVRIP active consumer",
      "medias": [
        "audio, sendonly, PCMA/8000, PCMU/8000"
      ],
      "senders": [
        "0 ANY, bytes=0, receivers=1"
      ]
    }
  ],
  "consumers": [
    {
      "type": "RTSP passive consumer",
      "url": "rtsp://127.0.0.1:8554/sala",
      "remote_addr": "127.0.0.1:45204",
      "user_agent": "FFmpeg Frigate/0.15.2-3bda638",
      "sdp": "v=0\r\no=- 1 1 IN IP4 0.0.0.0\r\ns=go2rtc/1.9.2\r\nc=IN IP4 0.0.0.0\r\nt=0 0\r\nm=video 0 RTP/AVP 96\r\na=rtpmap:96 H265/90000\r\na=fmtp:96 profile-id=1;sprop-vps=QAEMAf//AWAAAAMAsAAAAwAAAwCZrAwAAAMABAAAAwBSqA==;sprop-sps=QgEBAWAAAAMAsAAAAwAAAwCZoAFoIAZRY2u5MvzcBAQEEAAAAwAQAAADAUoQ;sprop-pps=RAHA8rA7NA==\r\na=control:trackID=0\r\n",
      "medias": [
        "video, sendonly, H264, H265",
        "audio, sendonly, MPEG4-GENERIC"
      ],
      "senders": [
        "96 H265, bytes=45118615, receivers=1"
      ],
      "send": 45588563
    },
    {
      "type": "probe",
      "remote_addr": "192.168.100.45:53367",
      "user_agent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/138.0.0.0 Safari/537.36",
      "medias": [
        "video, sendonly, ALL",
        "audio, sendonly, ALL",
        "audio, recvonly, ANY"
      ],
      "receivers": [
        "0 ANY, bytes=0, senders=1"
      ],
      "senders": [
        "255 H265, bytes=0, receivers=1",
        "0 PCMA/8000, bytes=0, receivers=1"
      ]
    }
  ]
}
```

**Comment by felipecrs** [comment]
Yes, 2-way audio only works through HTTPS.

I think you should test it with latest go2rtc version. You are using 1.9.2 which is relatively old.

https://docs.frigate.video/configuration/advanced#custom-go2rtc-version

**Comment by felipecrs** [comment]
Also, it would be nice if you tested it outside of Frigate to ensure the issue is really in go2rtc and not in Frigate.

**Comment by RomuloGatto** [comment]
By testing outside Frigate, if I expose go2rtc on a domain https covered, mic should show up at the stream?

**Comment by felipecrs** [comment]
Yes, when viewing it from:

```
https://your-go2rtc.example.com/webrtc.html?src=rua&media=video+audio+microphone
```

**Comment by fedd** [comment]
> Same behaviour

Check if you're editing the correct file :)

**Comment by RomuloGatto** [comment]
> Check if you're editing the correct file :)
I'm editting /config/config.yml, that's the one, right?



> Yes, when viewing it from:
> 
> ```
> https://your-go2rtc.example.com/webrtc.html?src=rua&media=video+audio+microphone
> ```

For testing purposes, I copied and adapted what @fedd did and tried to replicate, now it's like this:

```
go2rtc:
  streams:
    stairs_orig:
      - dvrip://admin:admin@192.168.100.30:34567?channel=0&subtype=0
    stairs:
      - "ffmpeg:stairs_orig#video=h264#audio=copy#hardware"
      - dvrip://admin:admin@192.168.100.30:34567?backchannel=1
    sala:
      - dvrip://admin:admin@192.168.100.30:34567
      - dvrip://admin:admin@192.168.100.30:34567?backchannel=1
```

And I noticed something weird. When I hit:

https://your-go2rtc.example.com/webrtc.html?src=stairs_orig&media=video+audio+microphone
https://your-go2rtc.example.com/webrtc.html?src=sala&media=video+audio+microphone
Asks for the microphone, but the video is stuck loading

https://your-go2rtc.example.com/webrtc.html?src=stairs&media=video+audio+microphone
Asks for the microphone, video loads, but I see no button to activate the mic. But then, when I did sounds on my computer, the camera played the sounds, so I assume it's good to go.

Now, just out of curiosity, why? 
Also, even if I point the frigate to this new `stairs`, the video gets stuck loading again

```
cameras:
  sala:
    ffmpeg:
      inputs:
        - path: rtsp://127.0.0.1:8554/stairs
          input_args: preset-rtsp-restream-low-latency
          roles:
            - detect
            - audio     
      output_args:
        record: preset-record-generic-audio-copy
```

**Comment by felipecrs** [comment]
Can you explain why this simple config is not good enough for you?

```yaml
go2rtc:
  streams:
    rua:
      - dvrip://admin:admin@192.168.1.22:34567
      - dvrip://admin:admin@192.168.1.22:34567?backchannel=1
```

**Comment by RomuloGatto** [comment]
Because mic works, but camera gets stuck at loading when hit https://your-go2rtc.example.com/webrtc.html?src=rua&media=video+audio+microphone.

The only endpoint that works camera and mic at same time is this:

https://your-go2rtc.example.com/webrtc.html?src=stairs&media=video+audio+microphone

    stairs:
      - "ffmpeg:stairs_orig#video=h264#audio=copy#hardware"
      - dvrip://admin:admin@192.168.100.30:34567?backchannel=1

**Comment by felipecrs** [comment]
What browser and version are you testing this in?

**Comment by RomuloGatto** [comment]
go2rtc 1.9.2
frigate 0.15.0

I'm running Chrome on macOS

I'll try to update go2rtc later today, just in case

**Comment by felipecrs** [comment]
Yeah, please update go2rtc for this test. But what's your Chrome version? 136+?

**Comment by RomuloGatto** [comment]
Yes, 138.0.7204.101 (arm64). And the behaviour is the same at Safari on iOS 18.5

**Comment by felipecrs** [comment]
Got it. If not mistaken, I think both should support H.265 in WebRTC. I hope updating go2rtc solves the issue, otherwise... it may be worth filling your own issue.

**Comment by RomuloGatto** [comment]
Well, it seems that everything was related to the go2rtc version. `rua` started working on the go2rtc preview URL, and on frigate as well! 
I guess, at least, we have it very well documented for the posteriority haha

@felipecrs, you are the one, man! Thank you so much!

**Comment by felipecrs** [comment]
Happy to help! :)

**Comment by khseal** [comment]
I use curl for testing. Some sounds stutter when converting. Maybe some ffmpeg parameters need to be adjusted?
Example command: curl -v -X POST   "http://ip:1984/api/streams?dst=tts_only&src=ffmpeg:./allowed.mp3%23audio=pcma%23input=file"
Here is an example sound file
[allowed.zip](https://github.com/user-attachments/files/21528430/allowed.zip)

**Comment by fklnc94** [comment]
im done on https frigate ui

---
go2rtc:
  streams:
    yol_main:
      - dvrip://tayu:wa4are@10.10.2.63:34567?backchannel=1
      - dvrip://tayu:wa4are@10.10.2.63:34567?channel=0&subtype=0
---
cameras:
  yol_sabit:
    ffmpeg:
      hwaccel_args: preset-intel-qsv-h264
      output_args:
        record: preset-record-generic-audio-copy
      inputs:
        - path: rtsp://127.0.0.1:8554/yol_main
          input_args: preset-rtsp-restream
          roles:
            - record
            - detect
    live:
      streams:
        Main: yol_main
---
